### PR TITLE
[SPARK-37013][CORE][SQL][FOLLOWUP] Use the new error framework to throw error in `FormatString` 

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This is a followup of https://github.com/apache/spark/pull/34313. The main change of this pr is  change to use the new error framework to throw error when `pattern.contains("%0$")` is true.

### Why are the changes needed?
Use the new error framework to throw error

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass the Jenkins or GitHub Action

### CODE REVIEW

**Comment by LuciferYang** [comment]
> let's use the new error framework to throw error in newly added code.

cc @cloud-fan ,  follow your advice in https://github.com/apache/spark/pull/34313

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 1]
```diff
@@ -1685,7 +1685,7 @@ case class ParseUrl(children: Seq[Expression], failOnError: Boolean = SQLConf.ge
 case class FormatString(children: Expression*) extends Expression with ImplicitCastInputTypes {
 
   require(children.nonEmpty, s"$prettyName() should take at least 1 argument")
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1687]: @cloud-fan this pr only change the  `newly added code`,  If also need to change line 1687 to use new error frame work too, please let me know

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -54,6 +54,9 @@
   "IF_PARTITION_NOT_EXISTS_UNSUPPORTED" : {
     "message" : [ "Cannot write, IF NOT EXISTS is not supported for table: %s" ]
   },
+  "ILLEGAL_FORMAT_ARGUMENT_INDEX" : {
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:57]: Shall we have a unified error class for illegal format string? cc @karenfeng

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

Gentle ping @karenfeng once more for @cloud-fan 's comment,  https://github.com/apache/spark/pull/34454/files#r740014071

**Comment by jerqi** [comment]
`attern.contains("%0$")` is true.
typo? 'attern' -> pattern?

**Comment by LuciferYang** [comment]
@jerqi thanks～already update the pr description

File: core/src/main/resources/error/error-classes.json (hunk 2 above)
Comment (karenfeng) [reply core/src/main/resources/error/error-classes.json:57]: There are a couple different ways for a string to have an illegal format, right? I'm not sure how to unify them all, but we can definitely have a more generic class here to represent that a string cannot contain a given substring, eg:

```
"ILLEGAL_SUBSTRING" : {
   "message" : [ "%s cannot contain %s." ]
}
```

Comment (LuciferYang) [reply core/src/main/resources/error/error-classes.json]: @karenfeng bf5790d fix this

**Comment by LuciferYang** [comment]
Anything else needs to be changed?

**Comment by LuciferYang** [comment]
Gentle ping @karenfeng once more

File: core/src/main/resources/error/error-classes.json (hunk 2 above)
Comment (karenfeng) [reply core/src/main/resources/error/error-classes.json]: @cloud-fan, does this LGTY? I don't have any more concerns.

Comment (LuciferYang) [reply core/src/main/resources/error/error-classes.json]: Gentle ping @cloud-fan @dongjoon-hyun

**Comment by cloud-fan** [comment]
Sorry I missed the ping. Merging to master, thanks!

**Comment by LuciferYang** [comment]
thanks all ~

**Pull request closed without merging**